scipy/optimize/_differentialevolution.py

```diff
@@ -265,6 +265,18 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
     >>> result.x, result.fun
     (array([1., 1., 1., 1., 1.]), 1.9216496320061384e-19)
 
+    If your objective function can be vectorized, then you may possibly
+    achieve speedups by the following:
```

Comment:
Sounds useful, but it's not immediately obvious to me how this works. Maybe we can discuss tomorrow?

Reply:
Sure.
The `workers` keyword can receive a map-like callable. The first argument is the function you want to map (the objective) and the second is an iterable (the parameter vectors). The map-like is supposed the return the objective function evaluated for each of th parameter vectors. Normally one would use `multiprocessing.Pool.map` for parallelisation here. All we're doing here is providing a different map-like, one which takes advantage of an objective that can be vectorised.

Reply:
I see. I think the trouble is that I am unfamiliar with the term "map" in this context, even after drafting gh-14334. It is also much more intuitive to pass a number to a keyword arg named `workers` than it is to pass a callable.

I don't remember - do `constraints` get passed in as `func`, too?

A note like this in the documentation would work, but have you thought about an argument `vectorized`?  If `vectorized is True` and the user does not pass a callable for `workers`, a `maplike_for_vectorized_func` could be provided automatically if we standardize the interface; e.g. a vectorized `func` must accept an array of shape `(popsize, n)` where `popsize` is blah blah and `n` is blah blah` and return an array of shape `(popsize,).

But there's no harm in starting with a doc adjustment. Can always add convenience parameter later.

Reply:
For `differential_evolution` workers can either be:

- a number, which is used to create a `multiprocessing.Pool` internally, with the relevant number of processes.
- a map-like callable.

The latter is intended to allow other kinds of parallelisation, e.g. `mpi4py` map-likes, [`schwimmbad.JoblibPool`, `schwimmbad.MPIPool`](https://schwimmbad.readthedocs.io/en/latest/examples/index.html#using-mpipool), etc.

At the moment parallelisation is only switched on if `workers != 1`. This also has the side-effect of changing `updating` from `'immediate'` to `'deferred'`. If there is a `vectorised` argument that side-effect would also have to apply.
If the user wants to vectorise, parallelisation would probably have to be turned off as the two are probably mutually exclusive. Which one would take precedence?

`constraints` get passed in as a sequence of `{NonLinearConstraint, LinearConstraint, Bounds}`

Reply:
> constraints get passed in as a sequence of {NonLinearConstraint, LinearConstraint, Bounds}

Sorry, what I was after is whether the constraints are evaluated in parallel? (I remember working on these with you, but I don't remember the details.) When I asked "do constraints get passed in as `func`, too?", I meant to use `func` in a generic sense - not necessarily representing the objective function. The documentation says that `workers` gets called like `workers(func, iterable)`, so it seems to me that `workers` _could_ be not only be used to parallelize the evaluation of the objective function but also parallelize the evaluation of the constraints. Do we do that?

For the rest of this post, I'm going to assume that there are no constraints.

> If the user wants to vectorise, parallelisation would probably have to be turned off as the two are probably mutually exclusive. Which one would take precedence?

We get to decide. 

It seems to me that if the user passes in a map-like for `workers`, then that map-like is responsible for divvying up the iterable - so the user is responsible for vectorization, if that's what they want to do. It's out of our control. So in that case, a `vectorized` parameter would be ignored.

However, if they specify an `int` for `workers`, we provide the map-like, right? In that case, I don't see why parallelization and vectorization are mutually exclusive. I don't personally know how to write that map-like offhand, but I imagine it's possible.

But again, since we get to decide, maybe we say initially that we only make use of `vectorized=True` if `workers=1`. We could leave the possibility of supporting `vectorized=Ture` with `workers` > 1 as a future enhancement.

Again, just a thought. We can start with a doc enhancement like you have if you prefer, but I think we need to be a bit more detailed about what the map-like needs to do.

Reply:
At the moment the constraints are evaluated [serially](https://github.com/scipy/scipy/blob/master/scipy/optimize/_differentialevolution.py#L1047). There's no real reason why they could use the same map-like.

Reply:
Was that "There's no real reason why they _couldn't_ use the same map-like."

Reply:
My experience is that constraint evaluation is typically negligible in time compared to objective function. Would you agree?

Reply:
> We could leave the possibility of supporting vectorized=Ture with workers > 1 as a future enhancement.

My initial thoughts on this are that if the user can make a vectorised objective function, they should also take on the responsibility of any parallelisation in such a situation (i.e. `vectorized and (workers > 1)` shouldn't be allowed).

Reply:
> My experience is that constraint evaluation is typically negligible in time compared to objective function. Would you agree?

My experience with trajectory optimization has been that constraints are a substantial portion of the computation, but I would only use gradient-based optimizers for that. So I'm not sure.